File: app/src/main/java/com/krystianwsul/checkme/domainmodel/TickData.java
package com.krystianwsul.checkme.domainmodel;

import android.content.Context;
import android.os.PowerManager;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import junit.framework.Assert;

import java.util.List;

public class TickData {
    private static final String WAKELOCK_TAG = "myWakelockTag";
    final boolean mSilent;

    @NonNull
    final String mSource;

    @NonNull
    final PowerManager.WakeLock mWakelock;

    @NonNull
    final List<Listener> listeners;

    public TickData(boolean silent, @NonNull String source, @NonNull Context context, @NonNull List<Listener> listeners) {
        Assert.assertTrue(!TextUtils.isEmpty(source));

        mSilent = silent;
        mSource = source;
        this.listeners = listeners;

        PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        Assert.assertTrue(powerManager != null);

        mWakelock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, WAKELOCK_TAG);
        mWakelock.acquire(30 * 1000);
    }

    void releaseWakelock() {
        if (mWakelock.isHeld()) mWakelock.release();
    }

    void release() {
        for (Listener listener : listeners)
            listener.onTick();

        releaseWakelock();
    }

    public interface Listener {

        void onTick();
    }
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/TickData.kt
package com.krystianwsul.checkme.domainmodel

import android.content.Context
import android.os.PowerManager
import com.krystianwsul.checkme.MyApplication
import com.krystianwsul.checkme.Preferences
import com.krystianwsul.checkme.domainmodel.notifications.Notifier
import com.krystianwsul.checkme.utils.time.toExactTimeStamp
import com.krystianwsul.common.time.ExactTimeStamp
import org.joda.time.DateTime

sealed class TickData {

    abstract val notifierParams: Notifier.Params

    abstract val shouldClear: Boolean

    abstract val waiting: Boolean

    abstract val domainChanged: Boolean

    abstract fun release()

    override fun toString() = super.toString() + ", notifierParams: $notifierParams"

    class Normal(override val notifierParams: Notifier.Params) : TickData() {

        override var shouldClear = false
            private set

        override val waiting = false

        override val domainChanged = false

        override fun release() {
            shouldClear = true
        }
    }

    class Lock(
        override val notifierParams: Notifier.Params,
        override val domainChanged: Boolean,
        val expires: ExactTimeStamp.Local = DateTime.now()
            .plusMillis(DURATION)
            .toExactTimeStamp(),
    ) : TickData() {

        init {
            Preferences.fcmLog.logLineHour("acquiring lock until $expires")
        }

        companion object {

            private const val WAKELOCK_TAG = "Check.me:myWakelockTag"

            private const val DURATION = 30 * 1000
        }

        private val wakelock = (MyApplication.instance.getSystemService(Context.POWER_SERVICE) as PowerManager).newWakeLock(
            PowerManager.PARTIAL_WAKE_LOCK,
            WAKELOCK_TAG
        )!!.apply {
            acquire(expires.long - ExactTimeStamp.Local.now.long)
        }

        override val shouldClear get() = expires < ExactTimeStamp.Local.now || !wakelock.isHeld

        override val waiting = true

        override fun release() {
            Preferences.fcmLog.logLineHour("releasing wakeLock, isHeld? " + wakelock.isHeld)

            if (wakelock.isHeld)
                wakelock.release()
        }

        override fun toString() = super.toString() + ", expires: $expires"
    }
}
